modules/modify-vm-template/pkg/templates/template-provider.go

```diff
@@ -0,0 +1,192 @@
+func (t *TemplateUpdator) ModifyTemplate() (*v1.Template, error) {
+	log.Logger().Debug("retrieving template", zap.String("name", t.cliOptions.GetTemplateName()), zap.String("namespace", t.cliOptions.GetTemplateNamespace()))
+	template, err := t.templateProvider.Get(t.cliOptions.GetTemplateNamespace(), t.cliOptions.GetTemplateName())
+	if err != nil {
+		return nil, err
+	}
+
+	updatedTemplate, err := t.UpdateTemplate(template)
+	if err != nil {
+		return nil, err
+	}
+
+	return t.templateProvider.Patch(updatedTemplate)
+}
+
+func (t *TemplateUpdator) UpdateTemplate(template *v1.Template) (*v1.Template, error) {
+	t.setValuesToTemplate(template)
+	vms, err := DecodeVM(template)
+	if err != nil {
+		return nil, err
+	}
+	updatedVM := t.setValuesToVMs(vms)
+
+	return EncodeVMToTemplate(template, updatedVM)
+}
+
+func (t *TemplateUpdator) setValuesToTemplate(template *v1.Template) {
+	labels := t.cliOptions.GetVMLabels()
+	if template.Labels == nil && len(labels) > 0 {
+		template.Labels = make(map[string]string)
+	}
+
+	for key, value := range labels {
+		template.Labels[key] = value
+	}
+
+	annotations := t.cliOptions.GetVMAnnotations()
+	if template.Annotations == nil && len(annotations) > 0 {
+		template.Annotations = make(map[string]string)
+	}
+
+	for key, value := range annotations {
+		template.Annotations[key] = value
+	}
+}
+
+func (t *TemplateUpdator) setValuesToVMs(vms []*kubevirtv1.VirtualMachine) []*kubevirtv1.VirtualMachine {
+	updatedVMs := make([]*kubevirtv1.VirtualMachine, len(vms))
+	labels := t.cliOptions.GetVMLabels()
+	annotations := t.cliOptions.GetVMAnnotations()
+
+	for i, vm := range vms {
+		if vm.Labels == nil && len(labels) > 0 {
+			vm.Labels = make(map[string]string)
+		}
+
+		for key, value := range labels {
+			vm.Labels[key] = value
+		}
+
+		if vm.Annotations == nil && len(annotations) > 0 {
+			vm.Annotations = make(map[string]string)
+		}
+
+		for key, value := range annotations {
+			vm.Annotations[key] = value
+		}
+
+		if vm.Spec.Template.Spec.Domain.CPU == nil {
+			vm.Spec.Template.Spec.Domain.CPU = &kubevirtv1.CPU{}
+		}
+
+		if sockets := t.cliOptions.GetCPUSockets(); sockets > 0 {
+			vm.Spec.Template.Spec.Domain.CPU.Sockets = uint32(sockets)
+		}
+		if cores := t.cliOptions.GetCPUCores(); cores > 0 {
+			vm.Spec.Template.Spec.Domain.CPU.Cores = uint32(cores)
+		}
+		if threads := t.cliOptions.GetCPUThreads(); threads > 0 {
+			vm.Spec.Template.Spec.Domain.CPU.Threads = uint32(threads)
+		}
+		if memory := t.cliOptions.GetMemory(); memory != nil {
+			vm.Spec.Template.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = *memory
+		}
+		updatedVMs[i] = vm
+	}
+
+	return updatedVMs
+}
+
+func EncodeVMToTemplate(template *templatev1.Template, vms []*kubevirtv1.VirtualMachine) (*v1.Template, error) {
+	vmsRaw := make([]runtime.RawExtension, len(vms))
+	for i, vm := range vms {
+		raw, err := json.Marshal(vm)
+		if err != nil {
+			return nil, err
+		}
+		vmsRaw[i].Raw = raw
+	}
+	template.Objects = vmsRaw
```

Comment:
I was not expecting to see multiple VMs in the template, but okay - better to be generic. Nevertheless there is also a chance other objects will be present and not just VMs. This statement would overwrite these objects